mllib/src/main/scala/org/apache/spark/mllib/stat/test/KSTest.scala

```diff
@@ -0,0 +1,191 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.mllib.stat.test
+
+import org.apache.commons.math3.distribution.{NormalDistribution, RealDistribution}
+import org.apache.commons.math3.stat.inference.KolmogorovSmirnovTest
+
+import org.apache.spark.rdd.RDD
+
+/**
+ * Conduct the two-sided Kolmogorov Smirnov test for data sampled from a
+ * continuous distribution. By comparing the largest difference between the empirical cumulative
+ * distribution of the sample data and the theoretical distribution we can provide a test for the
+ * the null hypothesis that the sample data comes from that theoretical distribution.
+ * For more information on KS Test: https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
+ *
+ * Implementation note: We seek to implement the KS test with a minimal number of distributed
+ * passes. We sort the RDD, and then perform the following operations on a per-partition basis:
+ * calculate an empirical cumulative distribution value for each observation, and a theoretical
+ * cumulative distribution value. We know the latter to be correct, while the former will be off by
+ * a constant (how large the constant is depends on how many values precede it in other partitions).
+ * However, given that this constant simply shifts the ECDF upwards, but doesn't change its shape,
+ * and furthermore, that constant is the same within a given partition, we can pick 2 values
+ * in each partition that can potentially resolve to the largest global distance. Namely, we
+ * pick the minimum distance and the maximum distance. Additionally, we keep track of how many
+ * elements are in each partition. Once these three values have been returned for every partition,
+ * we can collect and operate locally. Locally, we can now adjust each distance by the appropriate
+ * constant (the cumulative sum of # of elements in the prior partitions divided by the data set
+ * size). Finally, we take the maximum absolute value, and this is the statistic.
+ */
+private[stat] object KSTest {
+
+  // Null hypothesis for the type of KS test to be included in the result.
+  object NullHypothesis extends Enumeration {
+    type NullHypothesis = Value
+    val oneSampleTwoSided = Value("Sample follows theoretical distribution.")
+  }
+
+  /**
+   * Runs a KS test for 1 set of sample data, comparing it to a theoretical distribution
+   * @param data `RDD[Double]` data on which to run test
+   * @param cdf `Double => Double` function to calculate the theoretical CDF
+   * @return KSTestResult summarizing the test results (pval, statistic, and null hypothesis)
+   */
+  def testOneSample(data: RDD[Double], cdf: Double => Double): KSTestResult = {
+    val n = data.count().toDouble
+    val localData = data.sortBy(x => x).mapPartitions { part =>
+    val partDiffs = oneSampleDifferences(part, n, cdf) // local distances
```

Comment:
done. Sorry, I misunderstood the comment before and thought you meant to un-indent the entire thing, not just the last line. Fixed now